[WIP] Final refactoring of header names and add test script

chaithyagr opened the pull request:
This PR does final movements and renames as a part of issue #1675 

Also, I have added a minor python script that uses find and grep to look for cases of files that dont follow the rules as specified in #1675 

Currently, the only outliers (after running the code, and after this PR) are:
```
./cpp/src/decisiontree/memory.h:#include "memory.cuh"
./cpp/examples/dbscan/dbscan_example.cpp:#include <cuml/common/cubAllocatorAdapter.cuh>
./cpp/examples/kmeans/kmeans_example.cpp:#include <cuml/common/cubAllocatorAdapter.cuh>
```

The first is a decisiontree, and it needs to be refactored.
The other 2 actually come up as a part of rename of cubAllocatorAdapter.hpp to cubAllocatorAdapter.cuh as it includes #include <cub/util_allocator.cuh>

However, these includes are guarded by a ifdef, so they need not be a CUDA source file. 
@teju85 how do you think it is best to fix this? 

I can add these to ignore the list of files in the test script for now, so that the test can start running for now on newer PRs. 


A few comments on the script itself:

1) It's really basic find and grep, but it seems to do te job. I didnt want to use python related libaries to carry this out for that reason. Due to this, there is a disadvantage that it can run only on UNIX environment, however, since this is more of a code style check, I dont think it could be a problem

2) This script just ensures that there are no files that actually does not follow the rules as specified in #1675 . It does NOT check if all the other source files (esp CUDA source files), actually have device functions or not. I am not sure if we need that, and if we do, the checks would be a bit complicated, but we can get it also. @teju85 opinions?


I am making this as a draft as we may still need to finalize on steps we plan to take for this PR.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

teju85 commented:
@chaithyagr ouch... I didn't realize that, my bad! `cubAllocatorAdapter` should actually have been a `.hpp` itself :( because [util_allocator.cuh](https://github.com/thrust/cub/blob/master/cub/util_allocator.cuh) does not depend on anything cuda device/kernel calls. So, please rename this file back to `.hpp`. This means you might have to add special cases inside your script for such things.

chaithyagr commented:
> @chaithyagr ouch... I didn't realize that, my bad! `cubAllocatorAdapter` should actually have been a `.hpp` itself :( because [util_allocator.cuh](https://github.com/thrust/cub/blob/master/cub/util_allocator.cuh) does not depend on anything cuda device/kernel calls. So, please rename this file back to `.hpp`. This means you might have to add special cases inside your script for such things.

But that would mean util_allocator.cuh needs to be renamed to util_allocator.h am I right? I'm assuming this would have a more wide spread impact and hence is not done yet.

teju85 commented:
that's actually part of cub library. But since we know that it does not contain any device-related methods, we can make this "informed" decision.

chaithyagr commented on cpp/scripts/include_checker.py:
```diff
@@ -69,6 +69,7 @@ def check_includes_in(src):
 def main():
     args = parse_args()
     all_files = list_all_source_file(args.regex_compiled, args.dirs)
+    print(all_files)
```
Oops, this was a part of debug.

chaithyagr commented:
Makes sense. 
I'm opening this PR as ready to review. I have added a python script and taken in the above changes. 

One final thing is that recently some merges to master had `datasets/make_arima.hpp` which includes a cuh file. However, I am not sure if this hpp needs to move as well. For now, I have just added it to ignore in the script. @teju85 ?

Here is a regex of current exclusions:
 r"(cpp/src/decisiontree/memory.h$|cpp/include/cuml/common/cubAllocatorAdapter.hpp$|cpp/include/cuml/datasets/make_arima.hpp$)"


Finally, we need to make this script live on gpuCI, but I am unaware where to do this.

dantegd commented:
@chaithyagr @teju85 sorry for the late look at the PR, it seems mostly ready to merge, but now has some conflicts with branch-0.15, do you think you could solve the conflicts to go ahead with the PR?

chaithyagr commented:
Ahh, I will try to merge this to mainline tonight...

chaithyagr commented:
I finally could rebase to 823 commits, I hope the build is passing. I am running a local build too in meantime.

teju85 commented on cpp/include/cuml/datasets/make_arima.hpp:
```diff
@@ -16,8 +16,8 @@
 
-#include <cuml/tsa/arima_common.h>
 #include <cuml/cuml.hpp>
+#include <cuml/tsa/arima_common.cuh>
```
ah.. then this needs to be a .cuh too.

teju85 commented on cpp/scripts/check_extension_naming.py:
```diff
@@ -0,0 +1,93 @@
+                "cpp/src_prims",
+                "cpp/test"]
+
+IGNORE_FILES = r"(cpp/src/decisiontree/memory.h$|cpp/include/cuml/common/cubAllocatorAdapter.hpp$|cpp/include/cuml/datasets/make_arima.hpp$)"
```
I think it's better to remove the arima component from this regex, as per my above comment.

chaithyagr commented:
Where can I check the health of branch-0.15 ? I have seemed to be struggling to get a passing build, and I am not sure if it is a common issue. 
Also after merging, I see new conflicts which I will resolve (basically a couple of files have cuda_utils.cuh, but I am not sure if the include is needed.

GPUtester commented:
Can one of the admins verify this patch?

chaithyagr commented:
Reverting to WIP tag, as I dont seem to have a good passing build yet.

chaithyagr commented:
I have remade this PR over branch 0.17...
@teju85 perhaps you could have a quick final review of the changes. (Mostly all are direct).
Let us merge and close this and not get it lost. 
Removing WIP tag.

teju85 reviewed (changes requested):
Overall LGTM. Just a couple of minor comments.

chaithyagr commented on cpp/scripts/check-name-extentions.py:
```diff
@@ -0,0 +1,94 @@
+Search_Regex = re.compile(r"(#include.*.cuh|__device__|__global__)")
+
+def parse_args():
+    argparser = argparse.ArgumentParser("Runs clang-format on a project")
```
On it

chaithyagr commented on cpp/scripts/check-name-extentions.py:
```diff
@@ -0,0 +1,94 @@
+def parse_args():
+    argparser = argparse.ArgumentParser("Runs clang-format on a project")
+    argparser.add_argument("-regex", type=str,
+                           default=r"[.](h|hpp|cpp)$",
```
Yes, this is a disadvantage of current scheme. 
It can only check for positive violations, i.e. if we have C++ source files with CUDA. This doesnt check if a source file without any CUDA code is wrongly named .cu or .cuh. I am assuming this wont occur that often, esp after seeing the codebase. What do you think @teju85 ?
Personally, I dont mind adding checks for ensuring __device__/__global__ or *.cuh includes exist in all cu / cuh files, but I am not very sure if it includes all cases possible to define CUDA kernels

teju85 replied:
hmm.. that's a good point, especially the ability to use thrust/cub API's for calling into CUDA kernels! Let's park this thought, for now.
